### ISSUE TITLE
How to unit test CommandHandlers?

### ISSUE THREAD

**Issue by Megadesty** [eventflow/EventFlow#719]
I have a question regarding the testability of `CommandHandler`s. Let's assume I have the classes
```csharp
public class Car : AggregateRoot<Car, CarId> { ... }

public class StartEngine : Command<Car, CarId, IExecutionResult> { ... }
public class StartEngineHandler : CommandHandler<Car, CarId, IExecutionResult, StartEngine> { ... }
```
and I want to unit test the `StartEngineHandler`. One way could be:
```csharp
using (var resolver = this.InitializeEventFlowAndGetResolver()) {
  var bus = resolver.Resolve<ICommandBus>();
  var actual = await bus.PublishAsync(new StartEngine());
```
At this point I can check `actual`, e.g. with `actual.Should().BeOfType<SuccessExecutionResult>();`. But even if that's true, I want to check if the CommandHandler really called the aggregate. So I could do
```csharp
  var queryProcessor = resolver.Resolve<IQueryProcessor>();
  var car = await queryProcessor.ProcessAsync(...);
  car.Should().BeEquivalentTo(new Car(...));
}  
```
because if the read model gets updated, that means the corresponding event was published, which is done by the aggregate, hence the `CommandHandler` called the right thing.

But, that is more an integration than an unit test. So how can I unit-test the `CommandHandler`? Since everything in EventFlow has an interface, I tried to use that to mock the aggregate in the `CommandHandler`'s unit test like this:

```csharp
public interface ICar : IAggregateRoot<CarId> { ... }
public class Car : AggregateRoot<Car, CarId>, ICar {
    public Car(CarId id) : base(id) {}
    ...
}

public class StartEngine : Command<ICar, CarId, IExecutionResult> { ... }
public class StartEngineHandler : CommandHandler<ICar, CarId, IExecutionResult, StartEngine> { ... }
```
This way I can use `StartEngineHandler.ExecuteCommandAsync(new Mock<ICar>())` in my unit test and check afterwards, if the mock was called.

While the unit test is green, the app doesn't work anymore. The problem is, that this exception is thrown:
https://github.com/eventflow/EventFlow/blob/5694823e0b827a7997f05d2955118789bbe33c5a/Source/EventFlow/Aggregates/AggregateFactory.cs#L67-L70
because the interface (`TAggregate` is `ICar`) has obviously no constructor.

One idea could be to extend the method so that it also checks if `TAggregate` is an interface, if yes if there is exactly one derived type and count the constructors of that specific type. I tried it with reflection, but that won't work without breaking the support for `netstandard1.6`.

So another try was to use the container/resolver. I outlined what I mean by that in PullRequest #720. While that works and hopefully shows what I want to achieve, it's a bad hack, because I misuse the container/resolver to work around the reflection limitation. To let the code work I must setup the Id-class in the container, since `resolver.Resolve()` wants to initialize a new object (what I actually don't need at all) and therefore needs to know all parameter types of the constructor:
```csharp
serviceRegistration.Register<CarId>(c => new CarId("1"));
serviceRegistration.Register<ICar, Car>();
```
One improvement here would be if I could ask the resolver/container, what specific type is registered for an given interface, without having the resolver actually creating an instance of the class. But that still feels like a hack. As I said, reflection is no option before `netstandard2.0`.

So I would be very happy for any idea what I could do.

**Comment by ckoehler-pillar** [comment]
Here is an example of how I unit test my command handlers without getting into the realm of integration test or mocking. This is on the simple side.
```  
[Fact]
public async Task ApplyCommandToExampleAggregate()
{
	var exampleId = ExampleId.New;
	const string exampleType = "XX";          
	var exampleAggregate = new exampleAggregate(exampleId);
	//Aggregate gets generated by this previous command
	exampleAggregate.InitiateEventExample();

	var specifyExampleTypeCommand = new SpecifyExampleTypeCodeCommand(exampleId, exampleType);
	
	var handler = new SpecifyExampleTypeCodeCommandHandler();

	await handler.ExecuteAsync(exampleAggregate, specifyExampleTypeCommand, CancellationToken.None);

	exampleAggregate.UncommittedEvents.Should().HaveCount(2)
		.And.Subject.Last().AggregateEvent.Should().BeOfType<ExampleTypeCodeSpecified>();
}
```

**Comment by Megadesty** [comment]
@ckoehler-pillar Thanks for that idea. You need no EventFlow initialization, and the query/readmodel stuff is not involved, what is alot better than what happened in my integration test.

But although it's a big step in the right direction, it still has two flaws:
- The CommandHandler test must know details what's going on in the aggregate (that there will be a `DomainEvent` with the expected `AggregateEvent` in the `UncommittedEvents` list)
- Since the aggregate logic is used, it's still kind of an integration test

So I wonder if there is a way to allow CommandHandlers to be tested independent of the specific aggregate? Or if not, could that be a worthy feature to implement @rasmus ?

**Comment by ckoehler-pillar** [comment]
The job of command handlers in happy path is to hit an aggregate. You should verify this without mocking the aggregate as then you have to also mock state and arent testing the handlers responsibility. I have went down this route and it provided zero value compared to the approach above. Even if it comes at the cost of being 'still kind of an integration test'.

I like to verify that my command handler succeeds by either the execution result or the aggregate was modified. To call this integration is a bit pedantic when Command handlers implement `CommandHandler<ExampleAggregate, ExampleId, ExampleExecutionResult, ExampleCommand>` Although I see where you are coming from as my team has had these arguments frequently and have swapped our approach many times. 

One thing that I try hard to follow is a 'thin controller' style and have thin handlers and ship the heavy lifting logic outside so each piece can be tested independently. While having handlers verify execution result and aggregate. As It is easy to test, and the test are trust worthy.

If you REALLY want to you can use nsub and do something like var exampleAggregate = SubstitutePartsOf<ExampleAggregate>; and then setup the method your command handler will call and verify that it received a call. I don't suggest this and I'd just strongly suggest writing test against your aggregate and aggregate states and ensures all aggregate / state code you wrote is functional. 

I hope this helps and makes sense.

**Comment by rasmus** [comment]
"Testability" is something that's really important for me and if there's any part of the EventFlow that should be easier to test I think it merits an investigation.

At our department, the command handlers have very little logic, basically just invoking the aggregate and thus never tested. Any code that required unit testing (as opposed to domain testing, as the command handlers are always tested there) are typically in either the aggregate or in some domain service.

**Comment by Megadesty** [comment]
Thanks to both of you for sharing your thoughts!
